pkg/ruler/compat.go

```diff
@@ -22,17 +24,21 @@ type Pusher interface {
 // https://github.com/prometheus/prometheus/pull/2000#discussion_r79108319 for
 // reasons why.
 type appendableAppender struct {
+	sync.Mutex
 	pusher  Pusher
 	ctx     context.Context
 	labels  []labels.Labels
 	samples []client.Sample
+	userID  string
 }
 
 func (a *appendableAppender) Appender() (storage.Appender, error) {
 	return a, nil
 }
 
 func (a *appendableAppender) Add(l labels.Labels, t int64, v float64) (uint64, error) {
+	a.Lock()
```

Comment:
Why are we locking here? None of the Appender implementations in prometheus lock suggesting these functions are not reentrant.

Reply:
A lock is required here because we pool the samples for a user to the same appendable.

Reply:
I think previously in Cortex we created a separate `appendableAppender` for each group, which was run on a single goroutine.  So there must also be some change that means we have more goroutines talking to the same `appendableAppender` now?

Reply:
Yea the rule groups for each user will all share an appendableAppender. There are some advantages  to this approach. Primarily, it will make  #1396 easier to solve since output limits for a user can be configured in the same place.